crates/bindings/src/upper_case.rs

```diff
@@ -0,0 +1,82 @@
+use crate::service::{self, Channel, Service};
+
+/// UpperCase is an example Service that returns its inputs as upper-case
+/// ASCII, along with a running sum of the number of bytes upper-cased
+/// (returned with each response code).
+pub struct UpperCase {
+    sum_length: u32,
+}
+
+impl Service for UpperCase {
+    fn create() -> Self {
+        Self { sum_length: 0 }
+    }
+
+    fn invoke(
+        &mut self,
+        _code: u32,
+        data: &[u8],
+        arena: &mut Vec<u8>,
+        out: &mut Vec<service::Out>,
+    ) {
+        let begin = arena.len() as u32;
+        arena.extend(data.iter().map(u8::to_ascii_uppercase));
+
+        out.push(service::Out {
+            code: self.sum_length,
+            begin,
+            end: arena.len() as u32,
+        });
+
+        self.sum_length += data.len() as u32;
+    }
+}
+
+// Define cbindgen <=> CGO hooks for driving the UpperCase service.
+
+#[no_mangle]
```

Comment:
Might be more trouble than it's worth, but there's the [paste crate](https://github.com/dtolnay/paste) that allows you to template sets of functions like this using `macro_rules` macros. So you'd be able to do something like `expose_c_functions!{"upper_case", UpperCase}` to generate these. Macros can be tough to read/write, though, so fine if we'd rather just type it out the old fashioned way.

Reply:
I'd like to keep it simple for now. Can revisit later.